src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -2981,60 +3025,51 @@ def _do_construct_handler(self, node, parent):
-        if not ctrl or ctrl[0].items[0]:
+        # In fparser Loop_Control has 4 children, all but one should be None,
+        # the one which is not None defines the loop boundaries style:
+        # LoopCtrl(While_Loop, Counter_Loop, Concurrent_Loop, Concurrent_Loop)
+        if not ctrl or ctrl[0].items[0] is not None:
+            # do loops with no condition and do while loops
             annotation = ['was_unconditional'] if not ctrl else None
             loop = WhileLoop(parent=parent, annotations=annotation)
             loop.ast = node
             condition = [Fortran2003.Logical_Literal_Constant(".TRUE.")] \
                 if not ctrl else [ctrl[0].items[0]]
             self.process_nodes(parent=loop, nodes=condition)
+            # Create Loop body Schedule
             loop_body = Schedule(parent=loop)
+            loop_body.ast = node
             loop.addchild(loop_body)
-            self.process_nodes(parent=loop_body, nodes=node.content[1:-1])
-            return loop
-
-        # Second element of items member of Loop Control is itself a
-        # tuple containing: Loop variable, [start value expression,
-        # end value expression, step expression] Loop variable will be
-        # an instance of Fortran2003.Name and will be an integer.
-        loop_var = str(ctrl[0].items[1][0])
-        variable_name = str(loop_var)
-        data_symbol = _find_or_create_unresolved_symbol(
-            parent, variable_name, symbol_type=DataSymbol,
-            datatype=default_integer_type())
-        # The loop node is created with the _create_loop factory method as some
-        # APIs require a specialised loop node type.
-        loop = self._create_loop(parent, data_symbol)
-        loop.ast = node
-
-        # Get the loop limits. These are given in a list which is the second
-        # element of a tuple which is itself the second element of the items
-        # tuple:
-        # (None, (Name('jk'), [Int_Literal_Constant('1', None), Name('jpk'),
-        #                      Int_Literal_Constant('1', None)]), None)
-        limits_list = ctrl[0].items[1][1]
-
-        # Start expression child
-        self.process_nodes(parent=loop, nodes=[limits_list[0]])
-
-        # Stop expression child
-        self.process_nodes(parent=loop, nodes=[limits_list[1]])
+        elif ctrl[0].items[1] is not None:
+            # CounterLoops, its children are: Loop variable and Loop Limits
+            loop_var, limits_list = ctrl[0].items[1]
+            loop = self._create_bounded_loop(parent, loop_var, limits_list)
+            loop.ast = node
+            loop_body = loop.loop_body
+            loop_body.ast = node
+        elif ctrl[0].items[3] is not None:
+            triplet = walk(ctrl[0].items[3], Fortran2003.Forall_Triplet_Spec)
+            loop = None
+            for expr in triplet:
+                variable, start, stop, step = expr.items
+                new_loop = self._create_bounded_loop(parent, variable,
```

Comment:
Could/should we annotate this loop so that we know it was marked as having independent iterations? We might have to remove such an annotation if the loop body is subsequently altered though. Not essential for this PR.

Reply:
This is the reason that this is just a towards PR. An annotation feels wrong to me because it would not be true that we can ignore it (it needs to be invalidated), but maybe a local property. I added a TODO now to explore this.

Reply:
@arporter I am having second thoughts on this. The way concurrent is defined is that there can still be reductions and dependencies but the compiler is responsible of producing the correct code to respect these. So this annotation/property won't prevent the need to run our independent_iteratons/dependency analysis, because we still need to react to the dependencies or leave it serial. So, I propose not recording it and close the issue with this PR if you agree.

Reply:
Ah. Yes, in that case it's no different from any other loop as far as the PSyIR is concerned - it becomes a matter for the validation step of any transformation.

Reply:
Would it be worth keeping a note though as we might want to target those particular loops with transformations? It should only be used for information though - it wouldn't imply that the dependence analysis wasn't needed for instance.

Reply:
I think we should record that a loop has been declared as independent. It's something that @sergisiso and I disagree on - hopefully you're on my side @arporter :-)

Reply:
By the sounds of it, it's no different from ACC KERNELS - it seems to say to the compiler to try to parallelise it but it's up to the compiler to decide? I'll need to read the standard which I will do in a bit...

Reply:
"The DO CONCURRENT construct provides a means for the program to specify that individual loop
iterations have no interdependencies"

Reply:
> "The DO CONCURRENT construct provides a means for the program to specify that individual loop
> iterations have no interdependencies"

That doesn't seem to match what Sergi said: "The way concurrent is defined is that there can still be reductions and dependencies but the compiler is responsible of producing the correct code to respect these."?

Reply:
I've looked through the 2008 and 2018 specs and there is nothing about reductions and dependencies. The way the spec reads is that you are simply specifying that all iterations are independent. The example they give is an indirection where the user knows that the lookups are different for different iterations.

The only potentially relevant words I could find are the following but it is unclear to me if they rule out reductions but they do seem to rule out dependencies.

`A variable that is referenced in an iteration shall either be previously defined during that iteration, or shall not be defined or become undefined during any other iteration.`

Where did Sergi find his quote from? Perhaps the compiler vendors have gone further than the spec?

Reply:
For now I reverted it back to adding a TODO and keeping this PR a towards, since I only want to stop the parser breaking preventing any further processing.

We can explore this on the issue, but the only thing that the standard says is "The execution order of the iterations can be left indeterminate (DO CONCURRENT);" but is says nothing regarding if iterations can run in parallel or not. My understanding is that this is for the compiler to figure out. (there are some extra constraints F18 C1136 to C1141 that makes it easier for the compiler).
I believe that:
```
do concurrent (i=1:100)
    a = a  + array(i)
enddo
```
should not be parallelised without dealing with he a race condition. The standard just says that the order doesn't matter. (There is a default(shared) additional keyword that may make the loop independent, see `https://github.com/j3-fortran/fortran_proposals/issues/62` for a discussion on this)

Reply:
I would like to (not at all humbly) point out that the quote I gave before and repeat below is from the 2008 spec

> "The DO CONCURRENT construct provides a means for the program to specify that individual loop iterations have no interdependencies"

Reply:
I think we can leave this as a TODO for now and argue about it later :-)